File: exercism/kotlin/luhn/src/main/kotlin/Luhn.kt
class Luhn(val number: Long) {

  val addends = mutableListOf<Int>()

  val checkDigit: Int
    get() = addends.last()

  val checksum: Int
    get() = addends.sum()

  val isValid: Boolean
    get() = checksum % 10 == 0

  val create: Long
    get() = 10 * number + (10 - Luhn(10 * number).checksum % 10) % 10

  init {
    var x = number
    var i = 0
    if (x == 0L) addends += 0
    while (x > 0L) {
      var next = (x % 10).toInt()
      if (i++ % 2 > 0) next *= 2
      if (next > 9) next -= 9
      addends += next
      x /= 10
    }
    addends.reverse()
  }
}

File: exercism/elixir/luhn/luhn.exs
defmodule Luhn do
  @doc """
  Calculates the total checksum of a number
  """
  @spec checksum(String.t) :: integer
  def checksum(number) do
    number
    |> String.reverse
    |> String.graphemes
    |> Stream.with_index
    |> Stream.map(&compute_digit/1)
    |> Enum.sum
  end

  defp compute_digit({digit, index}) do
    x = String.to_integer(digit) * (rem(index, 2) + 1)
    x - 9 * div(x, 10)
  end

  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t) :: boolean
  def valid?(number) do
    rem(checksum(number), 10) == 0
  end

  @doc """
  Creates a valid number by adding the correct
  checksum digit to the end of the number
  """
  @spec create(String.t) :: String.t
  def create(number) do
    number <> to_string(9 - rem(checksum(number <> "9"), 10))
  end
end
